pageserver/src/tenant/layer_map.rs

```diff
@@ -47,14 +53,165 @@ pub struct LayerMap {
     pub frozen_layers: VecDeque<Arc<InMemoryLayer>>,
 
     /// All the historic layers are kept here
+    historic_layers: RTree<LayerEnvelope>,
 
-    /// TODO: This is a placeholder implementation of a data structure
-    /// to hold information about all the layer files on disk and in
-    /// S3. Currently, it's just a vector and all operations perform a
-    /// linear scan over it.  That obviously becomes slow as the
-    /// number of layers grows. I'm imagining that an R-tree or some
-    /// other 2D data structure would be the long-term solution here.
-    historic_layers: Vec<Arc<dyn Layer>>,
+    /// L0 layers has key range: (Key::MIN..Key::MAX) and locating them using R-Tree search is very inefficient.
+    /// So l) layers are also pushed in l0_layers vector.
+    l0_layers: Vec<Arc<dyn Layer>>,
+}
+
+struct LayerEnvelope {
+    layer: Arc<dyn Layer>,
+}
```

Comment:
rstar has a trait called `Envelope`. I'm pretty surprised LayerEnvelope doesn't implement that trait.

Reply:
From the `rstar` docs:

> An envelope defines how different bounding boxes of inserted children in an r-tree can interact, e.g. how they can be merged or intersected. This trait is not meant to be implemented by the user. Currently, only one implementation exists (`crate::AABB`) and should be used.

link: https://docs.rs/rstar/0.9.3/rstar/trait.Envelope.html

Reply:
LayerEnvelope implements `rstar::RTreeObject`.
I hope I have done everything according to rstar documentation.